pkg/metrics/sketch_series.go

```diff
@@ -78,6 +79,117 @@ func (sl SketchSeriesList) MarshalJSON() ([]byte, error) {
 	return reqBody.Bytes(), err
 }
 
+// MarshalSplitCompress uses the stream compressor to marshal and compress sketch series payloads.
+// If a compressed payload is larger than the max, a new payload will be generated. This method returns a slice of
+// compressed protobuf marshaled gogen.SketchPayload objects. gogen.SketchPayload is not directly marshaled - instead
+// it's contents are marshaled individually, packed with the appropriate protobuf metadata, and compressed in stream.
+// The resulting payloads (when decompressed) are binary equal to the result of marshaling the whole object at once.
+func (sl SketchSeriesList) MarshalSplitCompress() ([]*[]byte, error) {
+	input := bytes.NewBuffer(make([]byte, 0, 1024))
+	output := bytes.NewBuffer(make([]byte, 0, 1024))
+
+	// The Metadata field of gogen.SketchPayload is never written to - so pack an empty metadata as the footer
+	footer := []byte{0x12, 0}
+
+	compressor, e := stream.NewCompressor(input, output, []byte{}, footer, []byte{})
+	if e != nil {
+		return nil, e
+	}
+	payloads := []*[]byte{}
+	precompressionBuf := make([]byte, 1024)
+
+	dsl := make([]gogen.SketchPayload_Sketch_Dogsketch, 1)
+	for _, ss := range sl {
+		if len(ss.Points) > cap(dsl) {
+			dsl = make([]gogen.SketchPayload_Sketch_Dogsketch, len(ss.Points))
+		}
+
+		for i, p := range ss.Points {
+			b := p.Sketch.Basic
+			k, n := p.Sketch.Cols()
+			dsl[i] = gogen.SketchPayload_Sketch_Dogsketch{
+				Ts:  p.Ts,
+				Cnt: b.Cnt,
+				Min: b.Min,
+				Max: b.Max,
+				Avg: b.Avg,
+				Sum: b.Sum,
+				K:   k,
+				N:   n,
+			}
+		}
+
+		sketch := gogen.SketchPayload_Sketch{
+			Metric:      ss.Name,
+			Host:        ss.Host,
+			Tags:        ss.Tags,
+			Dogsketches: dsl[:len(ss.Points)],
+		}
+
+		// Pack the protobuf metadata
+		i := 0
+		precompressionBuf[i] = 0xa
+		i++
+		i = encodeVarintAgentPayload(precompressionBuf, i, uint64(sketch.Size()))
+
+		// Resize the pre-compression buffer if needed
+		totalItemSize := sketch.Size() + i
+		if totalItemSize > cap(precompressionBuf) {
+			precompressionBuf = append(precompressionBuf, make([]byte, totalItemSize-cap(precompressionBuf))...)
+			precompressionBuf = precompressionBuf[:cap(precompressionBuf)]
+		}
+
+		// Marshal the sketch to the precompression buffer
+		_, e := sketch.MarshalTo(precompressionBuf[i:])
+		if e != nil {
+			return nil, e
+		}
+
+		// Compress the protobuf metadata and the marshaled sketch
+		switch compressor.AddItem(precompressionBuf[:totalItemSize]) {
+		case stream.ErrItemTooBig, stream.ErrPayloadFull:
```

Comment:
Any reason to use a switch instead of a if ?

Reply:
Not really - when I started it had more case, but I simplified it over time. Will replace with an `if` :)